File: Data Structures & Algorithms/surrounded-regions/submission-2.cs
public class Solution {
    private List<(int dr, int dc)> _directions = [
        (1, 0),
        (0, 1),
        (0, -1),
        (-1, 0)
    ];

    public void Solve(char[][] board) {
        int nrows = board.Length;
        int ncols = board[0].Length;
        for (int row = 0; row < nrows; row++) {
            DFS(board, row, 0, nrows, ncols);
            DFS(board, row, ncols - 1, nrows, ncols);
        }

        for (int col = 0; col < ncols; col++) {
            DFS(board, 0, col, nrows, ncols);
            DFS(board, nrows - 1, col, nrows, ncols);
        }

        // if we haven't seen it, then set to 'X'
        for (int row = 0; row < nrows; row++) {
            for (int col = 0; col < ncols; col++) {
                if (board[row][col] == 'T')
                    board[row][col] = 'O';
                else if (board[row][col] == 'O')
                    board[row][col] = 'X';
            }
        }
    }

    public void DFS(char[][] board, int row, int col, int nrows, int ncols) {
        if (board[row][col] != 'O') {
            return;
        }

        board[row][col] = 'T';

        foreach (var direction in _directions) {
            var (nr, nc) = (direction.dr + row, direction.dc + col);
            if (nr < 0 || nr >= nrows || nc < 0 || nc >= ncols || board[nr][nc] != 'O') {
                continue;
            }

            DFS(board, nr, nc, nrows, ncols);
        }
    }
}

File: Data Structures & Algorithms/island-perimeter/submission-2.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        nrows, ncols = len(grid), len(grid[0])
        def dfs(row, col, grid, nrows, ncols):
            perimeter = 0
            if grid[row][col] == -1:
                return perimeter
            grid[row][col] = -1
            for dr, dc in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                nr, nc = row + dr, col + dc
                if 0 > nr or nr >= nrows:
                    perimeter += 1
                elif 0 > nc or nc >= ncols:
                    perimeter += 1
                elif grid[nr][nc] == 0:
                    perimeter += 1
                else:
                    perimeter += dfs(nr, nc, grid, nrows, ncols)
                
            return perimeter
                

                
        for row in range(nrows):
            for col in range(ncols):
                if grid[row][col] == 1:
                    return dfs(row, col, grid, nrows, ncols)
        return 0
    
